### PR TITLE
libnet: add support for custom interface names

### PR DESCRIPTION
**Pull request by akerouanton**
Fixes:

- https://github.com/moby/moby/issues/25181

**- What I did**

To support this a new netlabel is added: `com.docker.network.endpoint.ifname`.

It gives the ability to specify the interface name to be set by netdrivers when the interface is added / moved into the container's network namespace.

All builtin netdrivers support it.

**- How I did it**

First commit refactor `libnetwork/osl` to add a new field `dstPrefix`. So far, its field `dstName` was used to store both the name prefix and the auto-generated interface name. When an interface was created, `dstName` would take the prefix and later on that field would be mutated to take the final name. Now, the `dstPrefix` is stored separately and the `dstName` is the final interface name (either auto-generated, or statically defined).

The second commit is the actual feature implementation.

The third commit refactors how iface names are generated. We now scan all the interfaces tied to a `Namespace` to find the first gap in numeric suffixes. This allows users to statically define ifnames that match generated names. For instance, a user can set `eth2` for a specific endpoint, and libnetwork will use [0, 1] and [3, ...] for dynamic ifname assignment.

**- How to verify it**

- All netdrivers have a new integration test.
- A new integration test makes sure custom ifnames are preserved on live restore.
- Mixing static and dynamic ifname assignments with the same prefix is also tested through a new integration test.

```
$ docker run --rm -it --network=name=bridge,driver-opt=com.docker.network.endpoint.ifname=foobar alpine ip link show | grep foobar
11: foobar@if20: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
```

**- Description for the changelog**

```markdown changelog
- Add a new netlabel `com.docker.network.endpoint.ifname` to customize the interface name used when connecting a container to a network. It's supported by all built-in network drivers on Linux.
  - When a container is created with multiple networks specified, there's no guarantee on the order networks will be connected to the container. So, if a custom interface name uses the same prefix as the auto-generated names (eg. `eth`), the container might fail to start.
  - The recommended practice is to use a different prefix (eg. `en0`), or a numerical suffix high enough to never collide (eg. `eth100`).
  - This label can be specified on `docker network connect` via the `--driver-opt` flag, eg. `docker network connect --driver-opt=com.docker.network.endpoint.ifname=foobar …`.
  - Or via the long-form `--network` flag on `docker run`, eg. `docker run --network=name=bridge,driver-opt=com.docker.network.endpoint.ifname=foobar …`
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
CI failures look legit

File: integration/network/bridge/bridge_linux_test.go [hunk 1]
```diff
@@ -333,3 +334,21 @@ func TestPointToPoint(t *testing.T) {
...
+		ctr.WithEndpointSettings("bridge", &networktypes.EndpointSettings{
+			DriverOpts: map[string]string{
+				netlabel.Ifname: "foobar",
+			},
+		}))
+	defer ctr.Remove(ctx, t, apiClient, ctrID, containertypes.RemoveOptions{Force: true})
+
+	out, err := ctr.Output(ctx, apiClient, ctrID)
+	assert.NilError(t, err)
+	assert.Assert(t, strings.Contains(out.Stdout, ": foobar@if"), "expected ': foobar@if' in 'ip link show':\n%s", out.Stdout)
```
Comment (thaJeztah) [inline integration/network/bridge/bridge_linux_test.go:353]: Nit; if you use `is.Contains()` it should generate a failure that has both "expected" and "actual" output (although it _would_ escape the newlines);


```suggestion
	assert.Assert(t, is.Contains(stdout, ": foobar@if"))
```


e.g.;

```go
func TestContains(t *testing.T) {
	const stdout = `1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
3: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN qlen 1000
    link/gre 0.0.0.0 brd 0.0.0.0
4: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
5: erspan0@NONE: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
6: ip_vti0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
7: ip6_vti0@NONE: <NOARP> mtu 1428 qdisc noop state DOWN qlen 1000
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
8: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
9: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1000
    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
10: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN qlen 1000
    link/[823] 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
71: eth0@if72: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff`

	assert.Check(t, is.Contains(stdout, ": foobar@if"))

	assert.Check(t, strings.Contains(stdout, ": foobar@if"), "expected ': foobar@if' in 'ip link show':\n%s", stdout)
}
```

Produces;

```console
=== RUN   TestContains
    example_test.go:107: assertion failed: string "1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000\n    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00\n2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000\n    link/ipip 0.0.0.0 brd 0.0.0.0\n3: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN qlen 1000\n    link/gre 0.0.0.0 brd 0.0.0.0\n4: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN qlen 1000\n    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff\n5: erspan0@NONE: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN qlen 1000\n    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff\n6: ip_vti0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000\n    link/ipip 0.0.0.0 brd 0.0.0.0\n7: ip6_vti0@NONE: <NOARP> mtu 1428 qdisc noop state DOWN qlen 1000\n    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00\n8: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000\n    link/sit 0.0.0.0 brd 0.0.0.0\n9: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1000\n    link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00\n10: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN qlen 1000\n    link/[823] 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00\n71: eth0@if72: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP\n    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff" does not contain ": foobar@if"
    example_test.go:109: assertion failed: expression is false: strings.Contains(stdout, ": foobar@if"): expected ': foobar@if' in 'ip link show':
        1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
            link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
        2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
            link/ipip 0.0.0.0 brd 0.0.0.0
        3: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN qlen 1000
            link/gre 0.0.0.0 brd 0.0.0.0
        4: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN qlen 1000
            link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
        5: erspan0@NONE: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN qlen 1000
            link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
        6: ip_vti0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
            link/ipip 0.0.0.0 brd 0.0.0.0
        7: ip6_vti0@NONE: <NOARP> mtu 1428 qdisc noop state DOWN qlen 1000
            link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
        8: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN qlen 1000
            link/sit 0.0.0.0 brd 0.0.0.0
        9: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN qlen 1000
            link/tunnel6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
        10: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN qlen 1000
            link/[823] 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
        71: eth0@if72: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
            link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
--- FAIL: TestContains (0.00s)

```

File: integration/network/macvlan/macvlan_test.go [hunk 2]
```diff
@@ -640,3 +641,35 @@ func TestPointToPoint(t *testing.T) {
...
+
+func TestEndpointWithCustomIfname(t *testing.T) {
+	// verify the driver automatically provisions the 802.1q link (dm-dummy0.60)
+	skip.If(t, testEnv.IsRemoteDaemon)
+	skip.If(t, testEnv.IsRootless, "rootless mode has different view of network")
+
+	ctx := setupTest(t)
+	apiClient := testEnv.APIClient()
+
+	master := "dm-dummy0"
```
Comment (thaJeztah) [inline integration/network/macvlan/macvlan_test.go:653]: The comment at the top should probably go here, as this is where that part of the test happens?

```suggestion
	// verify the driver automatically provisions the 802.1q link (dm-dummy0.60)
	master := "dm-dummy0"
```

Comment (thaJeztah) [reply integration/network/macvlan/macvlan_test.go:653]: Also for fixed values, I started to use `const` more; I think it sometimes help glancing over a test to identify "dynamic" values (may change during the test) and fixtures;

```suggestion
	// verify the driver automatically provisions the 802.1q link (dm-dummy0.60)
	const master = "dm-dummy0"
```

File: integration/network/bridge/bridge_linux_test.go (hunk 1 above)
Comment (akerouanton) [reply integration/network/bridge/bridge_linux_test.go:353]: `is.Contains` generating an unreadable output is exactly why I didn't use it 🙂

File: integration/network/macvlan/macvlan_test.go (hunk 2 above)
Comment (akerouanton) [reply integration/network/macvlan/macvlan_test.go]: Ah, yeah I had these changes parked yesterday while I was trying to fix unobvious concurrency issues in the osl package. I should have put this PR back in draft mode.

I removed these comments as they don't say anything useful, and are even wrong. It was copy/pasted from other tests -- we should probably remove them in those other tests too.

File: libnetwork/osl/interface_linux.go [hunk 3]
```diff
@@ -232,33 +232,29 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
...
 	} else {
+		n.mu.Lock()
 		i.dstName = fmt.Sprintf("%s%d", dstPrefix, n.nextIfIndex[dstPrefix])
 		n.nextIfIndex[dstPrefix]++
+		n.mu.Unlock()
 	}
 
-	path := n.path
-	isDefault := n.isDefault
-	nlh := n.nlHandle
```
Comment (akerouanton) [inline libnetwork/osl/interface_linux.go:246]: Not sure if we're all going to agree on this code style.

I think that by not making copies of Namespace properties, we clearly signal that: 1. unaltered versions of Namespace properties are used; 2. they're safe to access concurrently.

File: integration/network/service_test.go [hunk 4]
```diff
@@ -465,3 +469,33 @@ func TestServiceWithDefaultAddressPoolInit(t *testing.T) {
...
+			},
+		}))
+	defer container.Remove(ctx, t, apiClient, ctrId, containertypes.RemoveOptions{Force: true})
+
+	d.Restart(t, "--live-restore=true")
+
+	res, err := container.Exec(ctx, apiClient, ctrId, []string{"ip", "-o", "link", "show", "foobar"})
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal(res.ExitCode, 0))
+	assert.Check(t, strings.Contains(res.Stdout(), ": foobar@if"), "expected ': foobar@if' in 'ip link show':\n%s", res.Stdout())
```
Comment (robmry) [inline integration/network/service_test.go:500]: Live-restore doesn't touch interfaces created before the daemon stopped (so the ip link show would work even if live-restore did nothing). But live-restore does need to reconstruct the `osl.Namespace`, `Interface.dstName` gets restored by netlink-listing the interfaces and matching IP addresses.

So, maybe it's worth trying to disconnect the network here, as it'll fail if it can't delete the container's interface (by name) - and/or something else that'll fail if the restore didn't work properly?

File: libnetwork/osl/interface_linux.go (hunk 3 above)
Comment (robmry) [reply libnetwork/osl/interface_linux.go:246]: Looks good - I think we can take it further (not in this PR), to end up with comments on struct members that clearly state which need mutex protection.

**Review by robmry — APPROVED** [review]
I don't think there's a test that shows what's expected to happen when an explicit interface name clashes with an already allocated name ... perhaps worth adding?

```
# docker network create n0
# docker run --rm -ti --network n0 --name c1 busybox
# docker network create n1
# docker network connect --driver-opt com.docker.network.endpoint.ifname=eth0 n1 c1
Error response from daemon: failed to add interface vethbdb03af to sandbox: error renaming interface "vethbdb03af" to "eth0": file exists
```

Because we don't have a "join order" for networks, a command like this will sometimes work, and sometimes won't ... just depending on which network connection happens to handled first.

```
# docker run --rm -ti --network name=n0,driver-opt=com.docker.network.endpoint.ifname=eth0 --network name=n1 --name c1 alpine
```

So, if I want the "n0" endpoint to be called "eth0" ... I either have to connect it first, then separately connect others (without specifying multiple networks in the `run`), or explicitly name all the interfaces, or use a different default prefix for the network. Maybe that's a limitation we need to document?

Perhaps it's also worth adding examples like the ones above of how to use the new endpoint option in the changelog note (or, just a mention that it goes in the `driver-opt` option)?

File: libnetwork/osl/interface_linux_test.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+func TestGenerateIfaceName(t *testing.T) {
+	testcases := []struct {
+		prefix string
+		names  []string
+		want   string
+	}{
+		{prefix: "test", names: []string{"test0", "test1"}, want: "test2"},
+		{prefix: "test", names: []string{"test0", "test2"}, want: "test1"},
+		{prefix: "test", names: []string{"test2"}, want: "test0"},
```
Comment (thaJeztah) [inline libnetwork/osl/interface_linux_test.go:27]: See my other comment; not sure how realistic it is to happen, but we should probably have test-cases with mixed prefixes, and one that has a `-` in the prefix; for example, this will fail, and pick `eth0`;


```suggestion
		{prefix: "test", names: []string{"test2"}, want: "test0"},
		{prefix: "eth", names: []string{"eth-0", "eth-1", "eth0"}, want: "eth1"},
```

File: libnetwork/drivers/bridge/bridge_linux.go [hunk 6]
```diff
@@ -1462,7 +1462,7 @@ func (d *driver) Join(ctx context.Context, nid, eid string, sboxKey string, jinf
 	if network.config.ContainerIfacePrefix != "" {
 		containerVethPrefix = network.config.ContainerIfacePrefix
 	}
-	if err := iNames.SetNames(endpoint.srcName, containerVethPrefix); err != nil {
+	if err := iNames.SetNames(endpoint.srcName, containerVethPrefix, netlabel.GetIfname(epOpts)); err != nil {
```
Comment (thaJeztah) [inline libnetwork/drivers/bridge/bridge_linux.go:1465]: Overall, I'm a bit on the fence on the approach; I know we have the `labels` as "fuzzy options", but making it part of the option-type seems to be digging in even deeper (`map[string]interface{}` being the API), and I'm wondering if we can have a more targeted approach (`WithInterfaceName(name)`, or `WithInterfaceName(prefix, name)` (if we want that option to _either_ set prefix _or_ name), where we parse the labels early on 🤔

**Review by thaJeztah — COMMENTED** [review]
Just leaving as comment; I tried digging in, but it's a lot of complexity and I'm wondering if we (also see some of my comments) can get this done without changing so many signatures (especially because it appears some of those already have the information available in some form or another)

File: libnetwork/osl/interface_linux.go [hunk 7]
```diff
@@ -217,11 +216,11 @@ func moveLink(ctx context.Context, nlhHost nlwrap.Handle, iface netlink.Link, i
...
-// AddInterface adds an existing Interface to the sandbox. The operation will rename
-// from the Interface SrcName to DstName as it moves, and reconfigure the
-// interface according to the specified settings. The caller is expected
-// to only provide a prefix for DstName. The AddInterface api will auto-generate
-// an appropriate suffix for the DstName to disambiguate.
+// AddInterface adds an existing Interface to the sandbox. The operation will
+// rename from the Interface SrcName to DstName as it moves, and reconfigure
+// the interface according to the specified settings. If dstPrefix is provided,
+// but not dstName, AddInterface will auto-generate a unique suffix and append
+// it to dstPrefix.
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/osl/interface_linux.go:227]: Woops looks like this comment slipped in the wrong commit - that's why you're getting confused by `dstName` being mentioned here.

I rewrote the comment to use present tense, but since this function is left untouched until the next (2nd) commit I rather removed that change from the 1st one.

> Some confusing logic in the code as well, but I probably need to look closer at that to understand what exactly it does, because it looks like for the default network, dstName is ignored?

Yes, this function is confusing... `isDefault` here isn't "the default network" but the "default network namespace" - in other words, the "host" netns / network. There might be good reasons for that but I don't know tbh.

File: libnetwork/endpoint.go [hunk 8]
```diff
@@ -536,7 +536,7 @@ func (ep *Endpoint) sbJoin(ctx context.Context, sb *Sandbox, options ...Endpoint
 		return fmt.Errorf("failed to get driver during join: %v", err)
 	}
 
-	if err := d.Join(ctx, nid, epid, sb.Key(), ep, sb.Labels()); err != nil {
+	if err := d.Join(ctx, nid, epid, sb.Key(), ep, ep.generic, sb.Labels()); err != nil {
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/endpoint.go:539]: > it looks like the information needed is all captured in the endpoint itself (which is passed and must implement the driverapi.JoinInfo interface; that interface already has a InterfaceNames() method (but not sure if that's related)

It definitely is related, but data flows the other way around. That is, netdrivers call `InterfacesNames().SetNames(srcName, dstPrefix, dstName)` on `driverapi.JoinInfo` to set the interface names.

Libnetwork's `sbJoin` method does the following:

- Call the netdriver `Join` method with the `Endpoint` passed as 2nd argument (implementor of `driverapi.JoinInfo`). The driver sets a bunch of values on the libnet's `EndpointInterface` through the setters defined in `driverapi.JoinInfo`.
- Then call `populateNetworkResources` which builds a list of `osl` options based on the values set on `EndpointInterface` at the previous step, and then call the `AddInterface()` method defined in the `osl` package with those options.

So, to implement your suggestion, I'd have to either break the unidirectional nature of `driverapi.JoinInfo`, or add yet another interface. I think both are equally bad and undesirable - and long term plan is to get rid of those interfaces.

> (I notice sandbox has a Labels() method, although I don't think it's currently part of an interface)

The custom interface name is an endpoint option, not a sandbox option.

File: libnetwork/netlabel/labels.go [hunk 9]
```diff
@@ -70,3 +73,21 @@ const (
...
+	opt, ok := opts[Ifname]
+	if !ok {
+		return ""
+	}
+	ifname, ok := opt.(string)
+	if !ok {
+		return ""
+	}
+	return ifname
+}
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/netlabel/labels.go:93]: Right 🤦 

As for inlining, I don't have strong opinions but I feel like having a dedicated accessor that takes care of casting the option somewhat hides the ugliness of these `map[string]interface{}` and make callers' code slightly easier to read.

**Comment by akerouanton** [comment]
Force pushing some changes to check if the integration test `TestInspectNetwork` fails on the CI since I see it failing locally but can't pinpoint the root cause. For the record, here's the only log that seems relevant during the test run:

```
time="2025-02-05T17:07:06.678119088Z" level=error msg="fatal task error" error="link /go/src/github.com/docker/docker/bundles/test-integration/TestInspectNetwork/d4a82747d8cf8/root/vfs/dir/881dc556aa2e8939b005e433f2cb2a79a8e60385c7b23c3ae5a5226991f70180-init/bin/[ /go/src/github.com/docker/docker/bundles/test-integration/TestInspectNetwork/d4a82747d8cf8/root/vfs/dir/881dc556aa2e8939b005e433f2cb2a79a8e60385c7b23c3ae5a5226991f70180-init/bin/adduser: file exists" module=node/agent/taskmanager node.id=lxewhfhiufgh3mpz9iam0ru9d service.id=sy93xyxaj46ikxtfvowh53whe task.id=tqlaop75asy7uzt3m5ajupryq
```

File: libnetwork/osl/namespace_linux.go [hunk 10]
```diff
@@ -402,7 +402,7 @@ func (n *Namespace) Destroy() error {
 func (n *Namespace) RestoreInterfaces(interfaces map[Iface][]IfaceOption) error {
 	// restore interfaces
 	for iface, opts := range interfaces {
-		i, err := newInterface(n, iface.SrcName, iface.DstPrefix, opts...)
+		i, err := newInterface(n, iface.SrcName, iface.DstPrefix, iface.DstName, opts...)
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/osl/namespace_linux.go:404]: > This commit doesn't compile, because newInterface signature was not modified (yet)?

Rebase issue - this should be fixed now.

> I'm actually wondering; dstName is an optional option, and because of that probably should not be a positional argument (which more commonly are used for required arguments); is there a reason we cannot pass interface name as part of opts ... ?

Both `dstPrefix` and `dstName` are now optional params - either one has to be provided.

File: libnetwork/osl/sandbox_linux_test.go [hunk 11]
```diff
@@ -381,7 +381,7 @@ func TestSandboxCreate(t *testing.T) {
 	}
 
 	for _, i := range tbox.Interfaces() {
-		err = s.AddInterface(context.Background(), i.SrcName(), i.DstName(),
+		err = s.AddInterface(context.Background(), i.SrcName(), i.dstPrefix, i.DstName(),
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/osl/sandbox_linux_test.go:384]: > These don't compile in this commit, because AddInterface signature was not (yet) changed

Rebase issue - this should be fixed now.

> It feels more organic to feed the interface-name into it when constructing (which could even be i.DstName() to take an argument, to enumerate suffixes), e.g. i.DstName(n.SuffixEnumerator) (where SuffixEnumerator would give the next suffix to use

`Namespace.RestoreInterfaces()`, which is the other function calling `newInterface`, does something wildly different with the output of `newInterface()`. That code is fragile IMO, so I don't think this PR is the right place to change what `newInterface()` does, and I think your suggestion for `SuffixEnumerator` would actually bring more unnecessary complexity.

File: libnetwork/osl/interface_linux.go [hunk 12]
```diff
@@ -333,19 +292,115 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
...
+
+func (n *Namespace) generateIfaceName(prefix string) string {
+	var suffixes []int
+	for _, i := range n.iFaces {
+		if strings.HasPrefix(i.DstName(), prefix) {
+			s := i.DstName()[len(prefix):]
+			if v, err := strconv.Atoi(s); err == nil {
+				suffixes = append(suffixes, v)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/osl/interface_linux.go]: Right, negative suffixes are now ignored.

File: libnetwork/osl/interface_linux.go [hunk 13]
```diff
@@ -333,19 +292,115 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
...
+		return prefix + "0"
+	}
+
+	sort.Ints(suffixes)
+
+	for i := 0; i < len(suffixes); i++ {
+		if i != suffixes[i] {
+			return prefix + strconv.Itoa(i)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/osl/interface_linux.go]: > This last bit of code looks to be to inject a suffix in the first possible slot; is it important to have consecutive numbers, or would it be fine to just pick "max + 1" ?

That could become an issue if "max + 1" overflows the maximum number of characters for interface names (IFNAMSIZ = 16 chars), eg. `prefix = 'mylonglongif' / highest suffix = '9999'`.

By looking for the lowest slot, this would be handled correctly (granted there's no 9999 networks attached to that container, in which case the issue isn't on our side).

I added a comment to clarify this.

> I was also playing to see if we could skip this part, which I think we could if we know the first loop already produced only consecutive numbers; basically, if the highest number matches the number of interfaces, I think we can tell "they must've been consecutive"

Actually that code doesn't test if there's no gaps in suffixes. Doing that would be slightly more involved, at which point I'm not sure this gives any benefit given the added complexity. I'd prefer to keep this code simple, stupid.

File: libnetwork/osl/interface_linux_test.go (hunk 5 above)
Comment (akerouanton) [reply libnetwork/osl/interface_linux_test.go]: Done - added a test case for negative numbers.

There was an edge case around `-0` being treated as `0` - it's now handled correctly.

File: libnetwork/drivers/bridge/bridge_linux.go (hunk 6 above)
Comment (akerouanton) [reply libnetwork/drivers/bridge/bridge_linux.go:1465]: I'm not sure to see where this hypothetical option would be placed / passed around as that doesn't seem to fit anywhere in the current design.

Basically, `epOpts` being a set of (mostly) user-defined options passed to the netdriver isn't that terrible since these are handled exclusively by the netdriver (or should be), and libnetwork should assume no knowledge about their semantics and how / if they're going to be handled.

If you're suggesting that this option is passed by libnetwork to the netdriver, then libnetwork has to know about that label and that would break that ^. Plus, it'd go against your suggestion to change less signatures.

If you're suggesting that it should be passed by the netdriver to libnetwork, that's probably overkill for passing just a single option. I'm not a big fan of coming up with a new design for the sake of a single option with no clear plans to reuse the same mechanism for other options. We can look into this in the future but I doubt it'd be used by anything else.

File: integration/network/service_test.go (hunk 4 above)
Comment (akerouanton) [reply integration/network/service_test.go:500]: Done, and also added a comment to clarify why the test calls `NetworkDisconnect`.

**Comment by akerouanton** [comment]
Okay, so the issue I was seeing locally with caused by 'Docker VMM'. Tests should be green now.

> I don't think there's a test that shows what's expected to happen when an explicit interface name clashes with an already allocated name ... perhaps worth adding?

Done.

> Perhaps it's also worth adding examples like the ones above of how to use the new endpoint option in the changelog note

Done. I added a recommendation regarding colliding prefixes.

File: libnetwork/osl/interface_linux.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply libnetwork/osl/interface_linux.go]: Gotcha! More stuff to untangle in our backlog 😂

File: libnetwork/endpoint.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply libnetwork/endpoint.go:539]: > The custom interface name is an endpoint option, not a sandbox option.

Ah, sorry, I meant "add a similar accessor to the `Endpoint` interface, like the `Sandbox` interface has, but as discussed in our call, we should consider all these internal, and have the flexibility of changing things / shuffle things around later.

File: libnetwork/netlabel/labels.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply libnetwork/netlabel/labels.go]: Yeah, it's fine for now to keep a Method / Function for this. We need to have a close look at all the fuzzy `genericopts` in a wider scope, and see if we can reduce the back-and-forth changing of strongly typed arguments to a generic one (and back! in some cases)

**Review by thaJeztah — APPROVED** [review]
LGTM after rebasing
